# Data Cleaning: Car Details Dataset

### The goal of this project is to inspect and clean the dataset in a suitable form for further exploratory data analysis and predictive modelling

## Importing Necessary Libraries

In [361]:
import numpy as np
import pandas as pd
from IPython.display import HTML, display

In [362]:
pd.options.display.float_format = "{:,.3f}".format

## Reading in the Dataset

In [363]:
path = r"/content/Car details v3.csv"

df = pd.read_csv(path)
display(HTML(f"<h3>Data Shape: {df.shape}</h3>"))
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm  5.000  
1       250Nm@ 1500-2500rpm  5.000  
2     12.7@ 2,700(kgm@ rpm)  5.000  
3  22.4 kgm at 1750-2750rpm  5.000  
4     11.5@ 4,500(kgm@ rpm)  5.000

In [364]:
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 5.0 MB


- Data has 8128 observations and 13 features
- Data utilizes about 5 mb of memory
- There're 4 numeric and 9 categorical columns
- `mileage`, `engine`, `max_power` and `torque` have `object` type, but they should be converted to numeric types
- Some of the columns have missing values

## Checking for Duplicates

In [365]:
df.duplicated().sum()

1202

In [366]:
df.duplicated(subset="name").sum()

6070

In [367]:
(df
 .loc[df.duplicated(subset=["name", "year"], keep=False)]
 .sort_values("name"))

name  year  selling_price  km_driven    fuel  \
7747                 Audi A4 2.0 TDI  2014        1500000      60000  Diesel   
906                  Audi A4 2.0 TDI  2014        1600000      44000  Diesel   
7374     Audi A4 35 TDI Premium Plus  2016        2450000      30000  Diesel   
5261     Audi A4 35 TDI Premium Plus  2016        1898999      46000  Diesel   
4951          Audi A6 35 TFSI Matrix  2019        5923000      11500  Petrol   
...                              ...   ...            ...        ...     ...   
3520  Volvo XC40 D4 Inscription BSIV  2019        3800000      20000  Diesel   
374   Volvo XC40 D4 Inscription BSIV  2019        3800000      20000  Diesel   
6213  Volvo XC40 D4 Inscription BSIV  2019        3800000      20000  Diesel   
3251          Volvo XC40 D4 R-Design  2018        3400000      22000  Diesel   
145           Volvo XC40 D4 R-Design  2018        3400000      22000  Diesel   

     seller_type transmission           owner     mileage   engine  \
7747  Individual    Automatic     First Owner  16.55 kmpl  1968 CC   
906   Individual    Automatic    Second Owner  16.55 kmpl  1968 CC   
7374      Dealer    Automatic     First Owner  18.25 kmpl  1968 CC   
5261      Dealer    Automatic     First Owner  18.25 kmpl  1968 CC   
4951      Dealer    Automatic  Test Drive Car  15.26 kmpl  1798 CC   
...          ...          ...             ...         ...      ...   
3520  Individual    Automatic     First Owner   18.0 kmpl  1969 CC   
374   Individual    Automatic     First Owner   18.0 kmpl  1969 CC   
6213  Individual    Automatic     First Owner   18.0 kmpl  1969 CC   
3251      Dealer    Automatic     First Owner   18.0 kmpl  1969 CC   
145       Dealer    Automatic     First Owner   18.0 kmpl  1969 CC   

       max_power               torque  seats  
7747  147.51 bhp  320Nm@ 1750-2500rpm  5.000  
906   147.51 bhp  320Nm@ 1750-2500rpm  5.000  
7374  187.74 bhp  400Nm@ 1750-3000rpm  5.000  
5261  187.74 bhp  400Nm@ 1750-3000rpm  5.000  
4951  187.74 bhp  320Nm@ 1400-4100rpm  5.000  
...          ...                  ...    ...  
3520     190 bhp                400Nm  5.000  
374      190 bhp                400Nm  5.000  
6213     190 bhp                400Nm  5.000  
3251     190 bhp                400Nm  5.000  
145      190 bhp                400Nm  5.000  

[5894 rows x 13 columns]

- There're 1202 definite duplicate entries
- There're some vehicles of the same name, but these differ w.r.t. other features and can be considered valid entries
- The 1202 duplicate entries should be deleted from the dataset

## Column-wise Inspection
 - There're some irregularities and inconsistencies within the columns
 - These issues will be inspected and handled accordingly

## Name

In [368]:
df.name

0             Maruti Swift Dzire VDI
1       Skoda Rapid 1.5 TDI Ambition
2           Honda City 2017-2020 EXi
3          Hyundai i20 Sportz Diesel
4             Maruti Swift VXI BSIII
                    ...             
8123               Hyundai i20 Magna
8124           Hyundai Verna CRDi SX
8125          Maruti Swift Dzire ZDi
8126                 Tata Indigo CR4
8127                 Tata Indigo CR4
Name: name, Length: 8128, dtype: object

In [369]:
df.name.value_counts()

Maruti Swift Dzire VDI                             129
Maruti Alto 800 LXI                                 82
Maruti Alto LXi                                     71
BMW X4 M Sport X xDrive20d                          62
Maruti Swift VDI                                    61
                                                  ... 
Skoda Fabia 1.4 TDI Ambiente                         1
Mahindra Scorpio VLX 2WD AT BSIII                    1
Renault KWID Climber 1.0 AMT                         1
Mahindra XUV300 W8 Option Dual Tone Diesel BSIV      1
Toyota Innova 2.5 GX (Diesel) 8 Seater BS IV         1
Name: name, Length: 2058, dtype: int64

In [370]:
(df
 .name
 .str.split(" ", n=2, expand=True)
 .set_axis(["company", "model", "edition"], axis=1))

company   model           edition
0      Maruti   Swift         Dzire VDI
1       Skoda   Rapid  1.5 TDI Ambition
2       Honda    City     2017-2020 EXi
3     Hyundai     i20     Sportz Diesel
4      Maruti   Swift         VXI BSIII
...       ...     ...               ...
8123  Hyundai     i20             Magna
8124  Hyundai   Verna           CRDi SX
8125   Maruti   Swift         Dzire ZDi
8126     Tata  Indigo               CR4
8127     Tata  Indigo               CR4

[8128 rows x 3 columns]

In [371]:
(df
 .pipe(lambda df_: pd.concat([df_
                              .name
                              .str.split(" ", n=2, expand=True)
                              .set_axis(["company", "model", "edition"], axis=1),
                              df_],
                             axis=1)))

company   model           edition                          name  year  \
0      Maruti   Swift         Dzire VDI        Maruti Swift Dzire VDI  2014   
1       Skoda   Rapid  1.5 TDI Ambition  Skoda Rapid 1.5 TDI Ambition  2014   
2       Honda    City     2017-2020 EXi      Honda City 2017-2020 EXi  2006   
3     Hyundai     i20     Sportz Diesel     Hyundai i20 Sportz Diesel  2010   
4      Maruti   Swift         VXI BSIII        Maruti Swift VXI BSIII  2007   
...       ...     ...               ...                           ...   ...   
8123  Hyundai     i20             Magna             Hyundai i20 Magna  2013   
8124  Hyundai   Verna           CRDi SX         Hyundai Verna CRDi SX  2007   
8125   Maruti   Swift         Dzire ZDi        Maruti Swift Dzire ZDi  2009   
8126     Tata  Indigo               CR4               Tata Indigo CR4  2013   
8127     Tata  Indigo               CR4               Tata Indigo CR4  2013   

      selling_price  km_driven    fuel seller_type transmission  \
0            450000     145500  Diesel  Individual       Manual   
1            370000     120000  Diesel  Individual       Manual   
2            158000     140000  Petrol  Individual       Manual   
3            225000     127000  Diesel  Individual       Manual   
4            130000     120000  Petrol  Individual       Manual   
...             ...        ...     ...         ...          ...   
8123         320000     110000  Petrol  Individual       Manual   
8124         135000     119000  Diesel  Individual       Manual   
8125         382000     120000  Diesel  Individual       Manual   
8126         290000      25000  Diesel  Individual       Manual   
8127         290000      25000  Diesel  Individual       Manual   

                     owner     mileage   engine   max_power  \
0              First Owner   23.4 kmpl  1248 CC      74 bhp   
1             Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2              Third Owner   17.7 kmpl  1497 CC      78 bhp   
3              First Owner   23.0 kmpl  1396 CC      90 bhp   
4              First Owner   16.1 kmpl  1298 CC    88.2 bhp   
...                    ...         ...      ...         ...   
8123           First Owner   18.5 kmpl  1197 CC   82.85 bhp   
8124  Fourth & Above Owner   16.8 kmpl  1493 CC     110 bhp   
8125           First Owner   19.3 kmpl  1248 CC    73.9 bhp   
8126           First Owner  23.57 kmpl  1396 CC      70 bhp   
8127           First Owner  23.57 kmpl  1396 CC      70 bhp   

                         torque  seats  
0                190Nm@ 2000rpm  5.000  
1           250Nm@ 1500-2500rpm  5.000  
2         12.7@ 2,700(kgm@ rpm)  5.000  
3      22.4 kgm at 1750-2750rpm  5.000  
4         11.5@ 4,500(kgm@ rpm)  5.000  
...                         ...    ...  
8123           113.7Nm@ 4000rpm  5.000  
8124  24@ 1,900-2,750(kgm@ rpm)  5.000  
8125             190Nm@ 2000rpm  5.000  
8126        140Nm@ 1800-3000rpm  5.000  
8127        140Nm@ 1800-3000rpm  5.000  

[8128 rows x 16 columns]

### Observations:
- `name` provides the complete name of each vehicle
- Contains period, paranthesis, Roman numerals and numbers
- Overall, the entries seem to be valid and accurate



### Steps:
- The column can be broken into 2 or 3 parts, namely `company`, `model` and `edition` for brevity
- These new columns could be useful for further analysis

## Year

In [372]:
df.year

0       2014
1       2014
2       2006
3       2010
4       2007
        ... 
8123    2013
8124    2007
8125    2009
8126    2013
8127    2013
Name: year, Length: 8128, dtype: int64

In [373]:
df.year.describe()

count   8,128.000
mean    2,013.804
std         4.044
min     1,983.000
25%     2,011.000
50%     2,015.000
75%     2,017.000
max     2,020.000
Name: year, dtype: float64

### Observations:
- The column is of the right datatype
- The values seem to be valid and accurate
- No further cleaning steps required

## Selling Price

In [374]:
df.selling_price

0       450000
1       370000
2       158000
3       225000
4       130000
         ...  
8123    320000
8124    135000
8125    382000
8126    290000
8127    290000
Name: selling_price, Length: 8128, dtype: int64

In [375]:
df.selling_price.describe()

count        8,128.000
mean       638,271.808
std        806,253.404
min         29,999.000
25%        254,999.000
50%        450,000.000
75%        675,000.000
max     10,000,000.000
Name: selling_price, dtype: float64

### Observations:
- The column is of the right datatype
- The values seem to be valid
- No further cleaning steps required
- This could be treated as the target feature for predictive modelling

## KM Driven

In [376]:
df.km_driven

0       145500
1       120000
2       140000
3       127000
4       120000
         ...  
8123    110000
8124    119000
8125    120000
8126     25000
8127     25000
Name: km_driven, Length: 8128, dtype: int64

In [377]:
df.km_driven.describe()

count       8,128.000
mean       69,819.511
std        56,550.555
min             1.000
25%        35,000.000
50%        60,000.000
75%        98,000.000
max     2,360,457.000
Name: km_driven, dtype: float64

### Observations:
- The column is of the right datatype
- The values seem to be valid
- No further cleaning steps required

## Fuel

In [378]:
df.fuel

0       Diesel
1       Diesel
2       Petrol
3       Diesel
4       Petrol
         ...  
8123    Petrol
8124    Diesel
8125    Diesel
8126    Diesel
8127    Diesel
Name: fuel, Length: 8128, dtype: object

In [379]:
df.fuel.memory_usage(deep=True)

511907

In [380]:
(df
 .fuel
 .astype("category")
 .memory_usage(deep=True))

8674

In [381]:
511907 / 8674

59.016255476135576

In [382]:
df.fuel.unique()

array(['Diesel', 'Petrol', 'LPG', 'CNG'], dtype=object)

In [383]:
(df
 .fuel
 .value_counts()
 .pipe(lambda ser: pd.concat([ser, df.fuel
                                     .value_counts(normalize=True)],
                             axis=1)))

fuel  fuel
Diesel  4402 0.542
Petrol  3631 0.447
CNG       57 0.007
LPG       38 0.005

In [384]:
(df
 .loc[df.fuel.isin(["Diesel", "Petrol"])]
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['kmpl', nan], dtype=object)

In [385]:
(df
 .loc[df.fuel.isin(["CNG", "LPG"])]
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['km/kg', nan], dtype=object)

### Observations:
- `fuel` has 4 unique values
- Occupies about 511,900 bytes of memory
- CNG and LPG account for only 0.7% and 0.5% of the total observations respectively
- Vehicles operating on CNG and LPG have their mileage units in `km\kg` whereas those running on Petrol and Diesel have mileage measured on `kmpl`

### Steps:
- The datatype could be converted to `category` for optimizing memory usage (about 60 times less) since the cardinality is very less
- CNG and LPG only account for 1.2% of total observations collectively, whether to drop these rows could be decided on further exploratory analysis

## Seller Type

In [386]:
df.seller_type

0       Individual
1       Individual
2       Individual
3       Individual
4       Individual
           ...    
8123    Individual
8124    Individual
8125    Individual
8126    Individual
8127    Individual
Name: seller_type, Length: 8128, dtype: object

In [387]:
df.seller_type.unique()

array(['Individual', 'Dealer', 'Trustmark Dealer'], dtype=object)

In [388]:
df.seller_type.value_counts()

Individual          6766
Dealer              1126
Trustmark Dealer     236
Name: seller_type, dtype: int64

In [389]:
(df
 .loc[df.seller_type == "Trustmark Dealer"]
 .head())

name  year  selling_price  km_driven  \
375                 Toyota Camry 2.5 Hybrid  2016        2000000      68089   
376                      Maruti Wagon R LXI  2013         225000      58343   
378                           Honda Jazz VX  2016         550000      56494   
379  Toyota Innova 2.5 VX (Diesel) 7 Seater  2013         750000      79328   
380                Maruti Swift AMT VVT VXI  2019         650000       5621   

       fuel       seller_type transmission         owner     mileage   engine  \
375  Petrol  Trustmark Dealer    Automatic   First Owner  19.16 kmpl  2494 CC   
376  Petrol  Trustmark Dealer       Manual   First Owner  21.79 kmpl   998 CC   
378  Petrol  Trustmark Dealer       Manual   First Owner   18.2 kmpl  1199 CC   
379  Diesel  Trustmark Dealer       Manual  Second Owner  12.99 kmpl  2494 CC   
380  Petrol  Trustmark Dealer    Automatic   First Owner   22.0 kmpl  1197 CC   

     max_power               torque  seats  
375  157.7 bhp       213Nm@ 4500rpm  5.000  
376  67.05 bhp        90Nm@ 3500rpm  5.000  
378   88.7 bhp       110Nm@ 4800rpm  5.000  
379  100.6 bhp  200Nm@ 1200-3600rpm  7.000  
380  81.80 bhp       113Nm@ 4200rpm  5.000

In [390]:
(df
 .loc[df.seller_type == "Trustmark Dealer"]
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['kmpl'], dtype=object)

In [391]:
(df
 .loc[df.seller_type == "Individual"]
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['kmpl', 'km/kg', nan], dtype=object)

In [392]:
(df
 .loc[df.seller_type == "Dealer"]
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['kmpl', nan, 'km/kg'], dtype=object)

In [393]:
(df
 .seller_type
 .memory_usage(deep=True))

541616

In [394]:
(df
 .seller_type
 .astype("category")
 .memory_usage(deep=True))

8567

In [395]:
541616 / 8567

63.2211976187697

### Observations:
- `seller_type` has 3 unique values
- Occupies about 541,600 bytes of memory
- Vehicles of `Trustmark Dealer` have their mileages measured in `kmpl`
- Cardinality is less

### Steps:
- The datatype could be converted to `category` for optimizing memory usage (about 63 times) since the cardinality is very less

## Transmission

In [396]:
df.transmission

0       Manual
1       Manual
2       Manual
3       Manual
4       Manual
         ...  
8123    Manual
8124    Manual
8125    Manual
8126    Manual
8127    Manual
Name: transmission, Length: 8128, dtype: object

In [397]:
df.transmission.unique()

array(['Manual', 'Automatic'], dtype=object)

In [398]:
df.transmission.value_counts()

Manual       7078
Automatic    1050
Name: transmission, dtype: int64

In [399]:
(df
 .transmission
 .memory_usage(deep=True))

515342

In [400]:
(df
 .transmission
 .astype("category")
 .memory_usage(deep=True))

8493

In [401]:
515342 / 8493

60.67844106911574

### Observations:
- `transmission` has 2 unique values
- Occupies about 515,342 bytes of memory
- Cardinality is less

### Steps:
- The datatype could be converted to `category` for optimizing memory usage (about 61 times less) since the cardinality is very less

## Owner

In [402]:
df.owner

0                First Owner
1               Second Owner
2                Third Owner
3                First Owner
4                First Owner
                ...         
8123             First Owner
8124    Fourth & Above Owner
8125             First Owner
8126             First Owner
8127             First Owner
Name: owner, Length: 8128, dtype: object

In [403]:
df.owner.unique()

array(['First Owner', 'Second Owner', 'Third Owner',
       'Fourth & Above Owner', 'Test Drive Car'], dtype=object)

In [404]:
df.owner.value_counts()

First Owner             5289
Second Owner            2105
Third Owner              555
Fourth & Above Owner     174
Test Drive Car             5
Name: owner, dtype: int64

In [405]:
(df
 .owner
 .value_counts()
 .pipe(lambda ser: pd.concat([ser, df
                                    .owner
                                    .value_counts(normalize=True)],
                             axis=1)))

owner  owner
First Owner            5289  0.651
Second Owner           2105  0.259
Third Owner             555  0.068
Fourth & Above Owner    174  0.021
Test Drive Car            5  0.001

In [406]:
(df
 .owner
 .str.replace(" Owner", "")
 .unique())

array(['First', 'Second', 'Third', 'Fourth & Above', 'Test Drive Car'],
      dtype=object)

In [407]:
(df
 .owner
 .memory_usage(deep=True))

556518

In [408]:
(df
 .owner
 .astype("category")
 .memory_usage(deep=True))

8781

In [409]:
556518 / 8781

63.377519644687396

### Observations:

- There're 5 unique values
- The cardinality is less
- This column occupies about 556,518 bytes of memory
- `Test Drive Car` accounts for only 0.1% of the total observations

### Steps:

- The word `Owner` could be stripped off from the categories as it seems redundant
- The datatype could be converted to `category` for optimizing memory usage (about 63 times less) since the cardinality is very less

## Mileage

In [410]:
df.mileage

0        23.4 kmpl
1       21.14 kmpl
2        17.7 kmpl
3        23.0 kmpl
4        16.1 kmpl
           ...    
8123     18.5 kmpl
8124     16.8 kmpl
8125     19.3 kmpl
8126    23.57 kmpl
8127    23.57 kmpl
Name: mileage, Length: 8128, dtype: object

In [411]:
(df
 .mileage
 .str.split(" ")
 .str[1]
 .unique())

array(['kmpl', 'km/kg', nan], dtype=object)

In [412]:
(df
 .mileage
 .str.split(" ")
 .str[0]
 .astype("float")
 .describe())

count   7,907.000
mean       19.419
std         4.037
min         0.000
25%        16.780
50%        19.300
75%        22.320
max        42.000
Name: mileage, dtype: float64

In [413]:
(df
 .dropna()
 .loc[lambda df_: df_.mileage.str.startswith("0")])

name  year  selling_price  \
644   Tata Indica Vista Aura Safire Anniversary Edition  2009         135000   
785                              Hyundai Santro Xing GL  2009         120000   
1649                             Hyundai Santro Xing GL  2008         105000   
1676                Mercedes-Benz M-Class ML 350 4Matic  2011        1700000   
2137                    Land Rover Freelander 2 TD4 HSE  2013        1650000   
2366                       Hyundai Santro Xing (Non-AC)  2010         110000   
2725                       Hyundai Santro Xing (Non-AC)  2013         184000   
4527                Mercedes-Benz M-Class ML 350 4Matic  2011        1700000   
5276                             Hyundai Santro Xing GL  2008         175000   
5843                        Volkswagen Polo GT TSI BSIV  2014         574000   
5846                        Volkswagen Polo GT TSI BSIV  2014         575000   
5900                     Mahindra Bolero Pik-Up FB 1.7T  2020         679000   
6534                             Hyundai Santro Xing GL  2010         150000   
6629                    Mahindra Bolero Pik-Up CBC 1.7T  2019         722000   
6824                             Hyundai Santro Xing GL  2011         150000   
7002                       Hyundai Santro Xing (Non-AC)  2010         110000   
7337                      Mercedes-Benz GLC 220d 4MATIC  2017        3300000   

      km_driven    fuel seller_type transmission                 owner  \
644       28900  Petrol  Individual       Manual          Second Owner   
785       90000  Petrol  Individual       Manual          Second Owner   
1649     128000  Petrol  Individual       Manual           First Owner   
1676     110000  Diesel  Individual    Automatic           Third Owner   
2137      64788  Diesel      Dealer    Automatic           First Owner   
2366      80000  Petrol  Individual       Manual          Second Owner   
2725      15000  Petrol  Individual       Manual           First Owner   
4527     110000  Diesel  Individual    Automatic           Third Owner   
5276      40000  Petrol  Individual       Manual           First Owner   
5843      28080  Petrol      Dealer    Automatic           First Owner   
5846      28100  Petrol      Dealer    Automatic           First Owner   
5900       5000  Diesel  Individual       Manual           First Owner   
6534     110000  Petrol  Individual       Manual           First Owner   
6629      80000  Diesel  Individual       Manual           First Owner   
6824      40000  Petrol  Individual       Manual  Fourth & Above Owner   
7002      80000  Petrol  Individual       Manual          Second Owner   
7337      60000  Diesel      Dealer    Automatic           First Owner   

       mileage   engine   max_power                torque  seats  
644   0.0 kmpl  1172 CC      65 bhp  9.8@ 3,000(kgm@ rpm)  5.000  
785   0.0 kmpl  1086 CC      62 bhp       96.1Nm@ 3000rpm  5.000  
1649  0.0 kmpl  1086 CC      62 bhp       96.1Nm@ 3000rpm  5.000  
1676  0.0 kmpl  2987 CC     165 bhp        510@ 1600-2400  5.000  
2137  0.0 kmpl  2179 CC     115 bhp      400 Nm /2000 rpm  5.000  
2366  0.0 kmpl  1086 CC    62.1 bhp       96.1Nm@ 3000rpm  5.000  
2725  0.0 kmpl  1086 CC    62.1 bhp       96.1Nm@ 3000rpm  5.000  
4527  0.0 kmpl  2987 CC     165 bhp        510@ 1600-2400  5.000  
5276  0.0 kmpl  1086 CC      62 bhp       96.1Nm@ 3000rpm  5.000  
5843  0.0 kmpl  1197 CC  103.25 bhp   175nm@ 1500-4100rpm  5.000  
5846  0.0 kmpl  1197 CC  103.25 bhp   175nm@ 1500-4100rpm  5.000  
5900  0.0 kmpl  2523 CC      70 bhp   200Nm@ 1400-2200rpm  2.000  
6534  0.0 kmpl  1086 CC      62 bhp       96.1Nm@ 3000rpm  5.000  
6629  0.0 kmpl  2523 CC      70 bhp   200Nm@ 1400-2200rpm  2.000  
6824  0.0 kmpl  1086 CC      62 bhp       96.1Nm@ 3000rpm  5.000  
7002  0.0 kmpl  1086 CC    62.1 bhp       96.1Nm@ 3000rpm  5.000  
7337  0.0 kmpl  1950 CC     194 bhp        400nm@ 2800rpm  5.000

In [414]:
(df
 .mileage
 .isna()
 .sum())

221

In [415]:
values = np.empty(len(df))

for i, value in df.mileage.items():
  try:
    splitted = value.split(" ")
  except AttributeError:
    values[i] = np.nan
  else:
    num = float(splitted[0])
    unit = splitted[1]
    if num == 0:
      new_num = np.nan
    elif unit == "kmpl":
      new_num = num * 2.35
    elif unit == "km/kg":
      new_num = num * 0.0016
    values[i] = new_num

pd.Series(values).rename("mileage_mpg").astype(np.float16)

0      55.000
1      49.688
2      41.594
3      54.062
4      37.844
        ...  
8123   43.469
8124   39.469
8125   45.344
8126   55.375
8127   55.375
Name: mileage_mpg, Length: 8128, dtype: float16

### Observations:

- This is a numeric column but it is of `object` type
- The values are measured in 2 different units, `kmpl` and `km/kg`
- Some vehicles show a reading of 0 mileage: (maybe because)
 - The actual mileage values weren't recorded for these vehicles
 - The odometer may have been reset to zero (this is illegal)

### Steps:

- We will extract the numeric values and store them as floats
- We will then convert all values to a common unit, i.e., `mpg`
 - 1 km/l = 2.35 mpg [reference](https://www.mpgtolitres.com/kml-to-mpg)
 - 1 km/kg = 0.0016 mpg [reference](https://math.stackexchange.com/questions/1141752/how-to-convert-kilometers-kilogram-km-kg-to-miles-gallon-mpg)
- The values showing zero mileage will be considered missing
 - Replace with `nan`

## Engine

In [416]:
df.engine

0       1248 CC
1       1498 CC
2       1497 CC
3       1396 CC
4       1298 CC
         ...   
8123    1197 CC
8124    1493 CC
8125    1248 CC
8126    1396 CC
8127    1396 CC
Name: engine, Length: 8128, dtype: object

In [417]:
df.engine.isna().sum()

221

In [418]:
df.loc[df.engine.isna()]

name  year  selling_price  km_driven  \
13                    Maruti Swift 1.3 VXi  2007         200000      80000   
31                      Fiat Palio 1.2 ELX  2003          70000      50000   
78                         Tata Indica DLS  2003          50000      70000   
87             Maruti Swift VDI BSIV W ABS  2015         475000      78000   
119                  Maruti Swift VDI BSIV  2010         300000     120000   
...                                    ...   ...            ...        ...   
7846                Toyota Qualis Fleet A3  2000         200000     100000   
7996             Hyundai Santro LS zipPlus  2000         140000      50000   
8009  Hyundai Santro Xing XS eRLX Euro III  2006         145000      80000   
8068             Ford Figo Aspire Facelift  2017         580000     165000   
8103                  Maruti Swift 1.3 VXi  2006         130000     100000   

        fuel seller_type transmission         owner mileage engine max_power  \
13    Petrol  Individual       Manual  Second Owner     NaN    NaN       NaN   
31    Petrol  Individual       Manual  Second Owner     NaN    NaN       NaN   
78    Diesel  Individual       Manual   First Owner     NaN    NaN       NaN   
87    Diesel      Dealer       Manual   First Owner     NaN    NaN       NaN   
119   Diesel  Individual       Manual  Second Owner     NaN    NaN       NaN   
...      ...         ...          ...           ...     ...    ...       ...   
7846  Diesel  Individual       Manual   First Owner     NaN    NaN       NaN   
7996  Petrol  Individual       Manual  Second Owner     NaN    NaN       NaN   
8009  Petrol  Individual       Manual  Second Owner     NaN    NaN       NaN   
8068  Diesel  Individual       Manual   First Owner     NaN    NaN       NaN   
8103  Petrol  Individual       Manual  Second Owner     NaN    NaN       NaN   

     torque  seats  
13      NaN    NaN  
31      NaN    NaN  
78      NaN    NaN  
87      NaN    NaN  
119     NaN    NaN  
...     ...    ...  
7846    NaN    NaN  
7996    NaN    NaN  
8009    NaN    NaN  
8068    NaN    NaN  
8103    NaN    NaN  

[221 rows x 13 columns]

In [419]:
df.loc[df.engine.isna()].isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           221
seats            221
dtype: int64

In [420]:
df.loc[df.mileage.isna()].isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           221
seats            221
dtype: int64

In [421]:
(df
 .engine
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser)))

0      1,248.000
1      1,498.000
2      1,497.000
3      1,396.000
4      1,298.000
          ...   
8123   1,197.000
8124   1,493.000
8125   1,248.000
8126   1,396.000
8127   1,396.000
Name: engine, Length: 8128, dtype: float64

In [422]:
(df
 .engine
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser))
 .describe())

count   7,907.000
mean    1,458.625
std       503.916
min       624.000
25%     1,197.000
50%     1,248.000
75%     1,582.000
max     3,604.000
Name: engine, dtype: float64

In [423]:
(df
 .engine
 .str.split(" ")
 .str[1]
 .unique())

array(['CC', nan], dtype=object)

### Observations:

- This should be a numeric column but is of `object` type
- The rows which are missing values for `engine`, also have values missing for:
 - `mileage`
 - `torque`
 - `max_power`
 - `seats`
- Upon inspection, these values seem to be missing due to not being recorded
 - Since these rows contain missing values in 5 columns, they could be deleted
 - Values could be imputed based on other features (multivariate imputation techniques)
 - Decision could be taken during modelling stage based on model performance
- Otherwise, the values seem to be valid

### Steps:

- The unit should be stripped from the values and attached to column name for better readability

## Max Power

In [424]:
df.max_power

0           74 bhp
1       103.52 bhp
2           78 bhp
3           90 bhp
4         88.2 bhp
           ...    
8123     82.85 bhp
8124       110 bhp
8125      73.9 bhp
8126        70 bhp
8127        70 bhp
Name: max_power, Length: 8128, dtype: object

In [425]:
(df
 .max_power
 .str.split(" ")
 .str[1]
 .unique())

array(['bhp', nan], dtype=object)

In [426]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser)))

0       74.000
1      103.520
2       78.000
3       90.000
4       88.200
         ...  
8123    82.850
8124   110.000
8125    73.900
8126    70.000
8127    70.000
Name: max_power, Length: 8128, dtype: float64

In [427]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser))
 .describe())

count   7,912.000
mean       91.518
std        35.822
min         0.000
25%        68.050
50%        82.000
75%       102.000
max       400.000
Name: max_power, dtype: float64

In [428]:
df.max_power.str.startswith("0").sum()

6

In [429]:
mask = np.zeros(len(df), dtype=bool)
for i, entry in df.max_power.items():
  try:
    splitted = entry.split(" ")
  except:
    mask[i] = False
  else:
    if entry.startswith("0"):
      mask[i] = True
    else:
      mask[i] = False
df[mask]

name  year  selling_price  km_driven    fuel  \
575               Maruti Alto K10 LXI  2011         204999      97500  Petrol   
576               Maruti Alto K10 LXI  2011         204999      97500  Petrol   
1442  Maruti Swift Dzire VDI Optional  2017         589000      41232  Diesel   
1443  Maruti Swift Dzire VDI Optional  2017         589000      41232  Diesel   
2549   Tata Indica Vista Quadrajet LS  2012         240000      70000  Diesel   
2550   Tata Indica Vista Quadrajet LS  2012         240000      70000  Diesel   

     seller_type transmission        owner mileage engine max_power torque  \
575   Individual       Manual  First Owner     NaN    NaN         0    NaN   
576   Individual       Manual  First Owner     NaN    NaN         0    NaN   
1442      Dealer       Manual  First Owner     NaN    NaN         0    NaN   
1443      Dealer       Manual  First Owner     NaN    NaN         0    NaN   
2549  Individual       Manual  First Owner     NaN    NaN         0    NaN   
2550  Individual       Manual  First Owner     NaN    NaN         0    NaN   

      seats  
575     NaN  
576     NaN  
1442    NaN  
1443    NaN  
2549    NaN  
2550    NaN

In [430]:
df.dropna().loc[lambda df_: df_.max_power.str.startswith("0")]

Empty DataFrame
Columns: [name, year, selling_price, km_driven, fuel, seller_type, transmission, owner, mileage, engine, max_power, torque, seats]
Index: []

In [431]:
df.max_power.isna().sum()

215

In [432]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser))
 .pipe(lambda ser: ser[ser == 0]))

575    0.000
576    0.000
1442   0.000
1443   0.000
2549   0.000
2550   0.000
Name: max_power, dtype: float64

In [433]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .replace("0", np.nan)
 .pipe(lambda ser: pd.to_numeric(ser))
 .pipe(lambda ser: df.loc[ser.isna()])
 .pipe(lambda df_: df_.loc[df_.mileage.notna()]))

name  year  selling_price  km_driven fuel seller_type  \
4933  Maruti Omni CNG  2000          80000     100000  CNG  Individual   

     transmission         owner     mileage  engine max_power torque  seats  
4933       Manual  Second Owner  10.9 km/kg  796 CC       bhp    NaN  8.000

In [434]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .pipe(lambda ser: pd.to_numeric(ser))
 .loc[4933])

nan

In [435]:
(df
 .max_power
 .str.split(" ")
 .str[0]
 .replace("0", np.nan)
 .pipe(lambda ser: pd.to_numeric(ser)))

0       74.000
1      103.520
2       78.000
3       90.000
4       88.200
         ...  
8123    82.850
8124   110.000
8125    73.900
8126    70.000
8127    70.000
Name: max_power, Length: 8128, dtype: float64

### Observations:

- This should be a numeric column but is of `object` type
- Values are measured in units of `brake horsepower (bhp)`
- Entry of index `4933` has an empty value
- Some entries have values as zero (invalid)

### Steps:

- The unit should be stripped from the values and attached to column name for better readability
- The values having 0 should be replaced as nan

## Torque

In [436]:
df.torque

0                  190Nm@ 2000rpm
1             250Nm@ 1500-2500rpm
2           12.7@ 2,700(kgm@ rpm)
3        22.4 kgm at 1750-2750rpm
4           11.5@ 4,500(kgm@ rpm)
                  ...            
8123             113.7Nm@ 4000rpm
8124    24@ 1,900-2,750(kgm@ rpm)
8125               190Nm@ 2000rpm
8126          140Nm@ 1800-3000rpm
8127          140Nm@ 1800-3000rpm
Name: torque, Length: 8128, dtype: object

In [437]:
(df
 .torque
 .str.lower()
 .unique())

array(['190nm@ 2000rpm', '250nm@ 1500-2500rpm', '12.7@ 2,700(kgm@ rpm)',
       '22.4 kgm at 1750-2750rpm', '11.5@ 4,500(kgm@ rpm)',
       '113.75nm@ 4000rpm', '7.8@ 4,500(kgm@ rpm)', '59nm@ 2500rpm',
       '170nm@ 1800-2400rpm', '160nm@ 2000rpm', '248nm@ 2250rpm',
       '78nm@ 4500rpm', nan, '84nm@ 3500rpm', '115nm@ 3500-3600rpm',
       '200nm@ 1750rpm', '62nm@ 3000rpm', '219.7nm@ 1500-2750rpm',
       '114nm@ 3500rpm', '115nm@ 4000rpm', '69nm@ 3500rpm',
       '172.5nm@ 1750rpm', '6.1kgm@ 3000rpm', '114.7nm@ 4000rpm',
       '60nm@ 3500rpm', '90nm@ 3500rpm', '151nm@ 4850rpm',
       '104nm@ 4000rpm', '320nm@ 1700-2700rpm', '250nm@ 1750-2500rpm',
       '145nm@ 4600rpm', '146nm@ 4800rpm', '343nm@ 1400-3400rpm',
       '200nm@ 1400-3400rpm', '200nm@ 1250-4000rpm',
       '400nm@ 2000-2500rpm', '138nm@ 4400rpm', '360nm@ 1200-3400rpm',
       '200nm@ 1200-3600rpm', '380nm@ 1750-2500rpm', '173nm@ 4000rpm',
       '400nm@ 1750-3000rpm', '400nm@ 1400-2800rpm',
       '200nm@ 1750-3000rp

In [438]:
(df
 .torque
 .str.extract(r"(^[0-9.]+).", expand=False)
 .pipe(lambda ser: pd.to_numeric(ser)))

0      190.000
1      250.000
2       12.700
3       22.400
4       11.500
         ...  
8123   113.700
8124    24.000
8125   190.000
8126   140.000
8127   140.000
Name: torque, Length: 8128, dtype: float64

In [439]:
df.torque.isna().sum()

222

In [440]:
(df
 .torque
 .str.extract(r"(^[0-9.]+).", expand=False)
 .pipe(lambda ser: pd.to_numeric(ser))
 .values)

array([190. , 250. ,  12.7, ..., 190. , 140. , 140. ])

In [441]:
values = (df
          .torque
          .str.extract(r"(^[0-9.]+).", expand=False)
          .pipe(lambda ser: pd.to_numeric(ser))
          .values)

for i, entry in (df
                 .torque
                 .str.lower()
                 .items()):
  try:
    splitted = entry.split(" ")
  except AttributeError:
    pass
  else:
    if "nm" in entry:
      pass
    elif "kg" in entry:
      values[i] = values[i] * 9.80665

(pd.Series(values)
   .rename("torque_nm")
   .pipe(lambda ser: pd.concat([df.torque, ser], axis=1)))

torque  torque_nm
0                190Nm@ 2000rpm    190.000
1           250Nm@ 1500-2500rpm    250.000
2         12.7@ 2,700(kgm@ rpm)    124.544
3      22.4 kgm at 1750-2750rpm    219.669
4         11.5@ 4,500(kgm@ rpm)    112.776
...                         ...        ...
8123           113.7Nm@ 4000rpm    113.700
8124  24@ 1,900-2,750(kgm@ rpm)    235.360
8125             190Nm@ 2000rpm    190.000
8126        140Nm@ 1800-3000rpm    140.000
8127        140Nm@ 1800-3000rpm    140.000

[8128 rows x 2 columns]

### Observations:

- This should be nuemric column but is of `object` type
- The values are listed mainly in 2 units: `kgm` and `Nm`
 - The units are very messy and inconsistent

### Steps:

- The unit should be stripped from the values and attached to column name for better readability
- The values will all be converted to a common unit of `Nm`
 - 1 kgm = 9.80665 Nm [reference](https://www.convertunits.com/from/kg-m/to/N-m)



## Seats

In [442]:
df.seats

0      5.000
1      5.000
2      5.000
3      5.000
4      5.000
        ... 
8123   5.000
8124   5.000
8125   5.000
8126   5.000
8127   5.000
Name: seats, Length: 8128, dtype: float64

In [443]:
df.seats.unique()

array([ 5.,  4., nan,  7.,  8.,  6.,  9., 10., 14.,  2.])

In [444]:
df.seats.value_counts()

5.000     6254
7.000     1120
8.000      236
4.000      133
9.000       80
6.000       62
10.000      19
2.000        2
14.000       1
Name: seats, dtype: int64

### Observations:

- This column seems to be valid
- Its of type float due to missing values
- No further cleaning required

## Data Cleaning Function

In [445]:
def clean_data(data):
  # this function takes in the dataset and returns the cleaned version

  def convert_mileage_mpg(ser):
    # this function takes in the mileage column
    # and returns all values in the unit of mpg

    new_values = np.empty(len(ser))
    for i, entry in ser.items():
      try:
        splitted = entry.split(" ")
      except AttributeError:
        new_values[i] = np.nan
      else:
        value = float(splitted[0])
        unit = splitted[1]
        if num == 0:
          new_values[i] = np.nan
        elif unit == "kmpl":
          new_values[i] = value * 2.35
        elif unit == "km/kg":
          new_values[i] = value * 0.0016
    return (pd.Series(new_values)
              .astype(np.float16))


  def convert_torque_nm(ser):
    # this function takes in the torque column
    # and returns all values in unit of Nm

    values = (ser
              .str.extract(r"(^[0-9.]+).", expand=False)
              .pipe(lambda ser: pd.to_numeric(ser))
              .values)
    for i, entry in (ser
                     .str.lower()
                     .items()):
      try:
        # to handle nan values
        entry.split(" ")
      except AttributeError:
        pass
      else:
        if "nm" in entry:
          pass
        elif "kg" in entry:
          values[i] = values[i] * 9.80665
    return pd.Series(values)


  return (data
          .drop_duplicates(ignore_index=True)
          .apply(lambda ser: ser.str.strip() if ser.dtype == "object" else ser)
          .pipe(lambda df_: pd.concat([df_
                                       .name
                                       .str.split(" ", n=2, expand=True)
                                       .set_axis(["company", "model", "edition"], axis=1),
                                       df_],
                                      axis=1))
          .assign(year=lambda df_: df_.year.astype(np.int16),
                  fuel=lambda df_: df_.fuel.astype("category"),
                  seller_type=lambda df_: df_.seller_type.astype("category"),
                  transmission=lambda df_: df_.transmission.astype("category"),
                  owner=lambda df_: df_.owner
                                       .str.replace(" Owner", "")
                                       .astype("category"),
                  mileage_mpg=lambda df_: convert_mileage_mpg(df_.mileage),
                  engine_cc=lambda df_: df_
                                         .engine
                                         .str.split(" ")
                                         .str[0]
                                         .pipe(lambda ser: pd.to_numeric(ser)),
                  max_power_bhp=lambda df_: df_
                                             .max_power
                                             .str.split(" ")
                                             .str[0]
                                             .replace("0", np.nan)
                                             .pipe(lambda ser: pd.to_numeric(ser,
                                                                             errors="coerce")),
                  torque_nm=lambda df_: convert_torque_nm(df_.torque))
          .drop(columns=["mileage", "engine", "max_power", "torque"])
          .reindex(columns=["name",
                            "company",
                            "model",
                            "edition",
                            "year",
                            "owner",
                            "fuel",
                            "seller_type",
                            "transmission",
                            "km_driven",
                            "mileage_mpg",
                            "engine_cc",
                            "max_power_bhp",
                            "torque_nm",
                            "seats",
                            "selling_price"]))

## Cleaned Data

In [446]:
df_cleaned = clean_data(df)
df_cleaned

name  company  model               edition  \
0                Maruti Swift Dzire VDI   Maruti  Swift             Dzire VDI   
1          Skoda Rapid 1.5 TDI Ambition    Skoda  Rapid      1.5 TDI Ambition   
2              Honda City 2017-2020 EXi    Honda   City         2017-2020 EXi   
3             Hyundai i20 Sportz Diesel  Hyundai    i20         Sportz Diesel   
4                Maruti Swift VXI BSIII   Maruti  Swift             VXI BSIII   
...                                 ...      ...    ...                   ...   
6921  Maruti Wagon R VXI BS IV with ABS   Maruti  Wagon  R VXI BS IV with ABS   
6922         Hyundai i20 Magna 1.4 CRDi  Hyundai    i20        Magna 1.4 CRDi   
6923                  Hyundai i20 Magna  Hyundai    i20                 Magna   
6924              Hyundai Verna CRDi SX  Hyundai  Verna               CRDi SX   
6925             Maruti Swift Dzire ZDi   Maruti  Swift             Dzire ZDi   

      year           owner    fuel seller_type transmission  km_driven  \
0     2014           First  Diesel  Individual       Manual     145500   
1     2014          Second  Diesel  Individual       Manual     120000   
2     2006           Third  Petrol  Individual       Manual     140000   
3     2010           First  Diesel  Individual       Manual     127000   
4     2007           First  Petrol  Individual       Manual     120000   
...    ...             ...     ...         ...          ...        ...   
6921  2013          Second  Petrol  Individual       Manual      50000   
6922  2014          Second  Diesel  Individual       Manual      80000   
6923  2013           First  Petrol  Individual       Manual     110000   
6924  2007  Fourth & Above  Diesel  Individual       Manual     119000   
6925  2009           First  Diesel  Individual       Manual     120000   

      mileage_mpg  engine_cc  max_power_bhp  torque_nm  seats  selling_price  
0          55.000  1,248.000         74.000    190.000  5.000         450000  
1          49.688  1,498.000        103.520    250.000  5.000         370000  
2          41.594  1,497.000         78.000    124.544  5.000         158000  
3          54.062  1,396.000         90.000    219.669  5.000         225000  
4          37.844  1,298.000         88.200    112.776  5.000         130000  
...           ...        ...            ...        ...    ...            ...  
6921       44.406    998.000         67.100     90.000  5.000         260000  
6922       52.969  1,396.000         88.730    219.700  5.000         475000  
6923       43.469  1,197.000         82.850    113.700  5.000         320000  
6924       39.469  1,493.000        110.000    235.360  5.000         135000  
6925       45.344  1,248.000         73.900    190.000  5.000         382000  

[6926 rows x 16 columns]

## Memory Comparison

In [447]:
df.memory_usage(deep=True).sum()

5202999

In [448]:
df_cleaned.memory_usage(deep=True).sum()

2302779

In [449]:
5202999 / 2302779

2.2594434811156434

## Final Remarks:

- The dataset is now cleaned and ready for further exploratory data analysis
- The cleaned dataset utilizes nearly 2.25 times less memory than the original dataset
- The `name` column was split into 3 parts:
 - Further analysis should be done if it needs to be split into 2 or 3 parts for better model performance
- The columns measuring `mileage`, `max_power`, `torque`, `seats` and `engine` have values missing in the same rows
 - These rows could be deleted, or
 - Imputed based on other features (depending on model performance)
- Some columns contain very rare categories (<1% of total observations)
 - These categories should be handled appropriately